test/e2e/resource_usage_gatherer.go

```diff
@@ -75,43 +70,193 @@ func (s *ResourceUsageSummary) PrintJSON() string {
+			dataMap[name].memWorkSetData = append(dataMap[name].memWorkSetData, data.MemoryWorkingSetInBytes)
+		}
+	}
+	for _, v := range dataMap {
+		sort.Float64s(v.cpuData)
+		sort.Sort(uint64arr(v.memUseData))
+		sort.Sort(uint64arr(v.memWorkSetData))
+	}
+
+	result := make(map[int]resourceUsagePerContainer)
+	for _, perc := range percentilesToCompute {
+		data := make(resourceUsagePerContainer)
+		for k, v := range dataMap {
+			percentileIndex := int(math.Ceil(float64(len(v.cpuData)*perc)/100)) - 1
+			data[k] = &containerResourceUsage{
+				Name:                    k,
+				CPUUsageInCores:         v.cpuData[percentileIndex],
+				MemoryUsageInBytes:      v.memUseData[percentileIndex],
+				MemoryWorkingSetInBytes: v.memWorkSetData[percentileIndex],
+			}
+		}
+		result[perc] = data
+	}
+	return result
+}
+
+func leftMergeData(left, right map[int]resourceUsagePerContainer) map[int]resourceUsagePerContainer {
+	result := make(map[int]resourceUsagePerContainer)
+	for percentile, data := range left {
+		result[percentile] = data
+		if _, ok := right[percentile]; !ok {
+			continue
+		}
+		for k, v := range right[percentile] {
+			result[percentile][k] = v
+		}
+	}
+	return result
+}
+
+type resourceGatherWorker struct {
+	c                    *client.Client
+	nodeName             string
+	wg                   *sync.WaitGroup
+	containerIDToNameMap map[string]string
+	containerIDs         []string
+	stopCh               chan struct{}
+	dataSeries           []resourceUsagePerContainer
+}
+
+func (w *resourceGatherWorker) singleProbe(errors []error) {
+	data := make(resourceUsagePerContainer)
+	nodeUsage, err := getOneTimeResourceUsageOnNode(w.c, w.nodeName, 15*time.Second, func() []string { return w.containerIDs }, true)
+	if err != nil {
+		errors = append(errors, err)
+		return
+	}
+	for k, v := range nodeUsage {
+		data[w.containerIDToNameMap[k]] = v
+	}
+	w.dataSeries = append(w.dataSeries, data)
+}
+
+func (w *resourceGatherWorker) gather(initialSleep time.Duration, errors []error) {
+	defer utilruntime.HandleCrash()
+	defer w.wg.Done()
+	select {
+	case <-time.After(initialSleep):
+		w.singleProbe(errors)
 		for {
 			select {
-			case <-time.After(period):
-				now := time.Now()
-				data, err := g.getKubeSystemContainersResourceUsage(c)
-				if err != nil {
-					Logf("Error while getting resource usage: %v", err)
-					continue
-				}
-				g.usageTimeseries[now] = data
-			case <-g.stopCh:
-				Logf("Stop channel is closed. Stopping gatherer.")
-				return nil
+			case <-time.After(resourceDataGatheringPeriod):
+				w.singleProbe(errors)
+			case <-w.stopCh:
+				return
 			}
 		}
+	case <-w.stopCh:
+		return
+	}
+}
+
+func (g *containerResourceGatherer) getKubeSystemContainersResourceUsage(c *client.Client) error {
+	errors := make([]error, 0)
+	delay := resourceDataGatheringPeriod / time.Duration(len(g.workers))
+	go func() {
+		for i := range g.workers {
+			go g.workers[i].gather(delay, errors)
```

Comment:
Passing "errors" here, makes it thread-unsafe.
In fact generally it seems that at the end of the day you are ignoring those errors anyway.

So what i would suggest is NOT passing "errors" at all to "gather" method, and only log those errors where they appear.